pgxn/neon/file_cache.c

```diff
@@ -432,10 +486,19 @@ lfc_read(NRelFileInfo rinfo, ForkNumber forkNum, BlockNumber blkno,
 	hash = get_hash_value(lfc_hash, &tag);
 
 	LWLockAcquire(lfc_lock, LW_EXCLUSIVE);
+
+	if (!LFC_ENABLED())
+	{
+		LWLockRelease(lfc_lock);
+		return false;
+	}
+
 	entry = hash_search_with_hash_value(lfc_hash, &tag, hash, HASH_FIND, NULL);
 	if (entry == NULL || (entry->bitmap[chunk_offs >> 5] & (1 << (chunk_offs & 31))) == 0)
 	{
 		/* Page is not cached */
+		lfc_ctl->misses += 1; /* race condition here, but precise value is not needed */
+		pgBufferUsage.file_cache.misses += 1;
```

Comment:
Well, I understand you point.
I also want to reduce number of changes in core as much as possible.
But both prefetch and LFC seems to have critical impact on Neon performance (certainly in addition to shared buffer access stats). So I think it is very important to present this information to user and in most convenient way.

Let's ask opinion of other people? @hlinnaka @tristan957  ?

Reply:
To me, it seems useful, though I do echo what Matthias says. Here is my comment in the PG PR.

> Has Postgres ever thought about letting extensions add to EXPLAIN? Seems like that would be useful.

I am wondering if having a `neon_stat_statements` table would be good, similar to [`pg_stat_statements`](https://www.postgresql.org/docs/current/pgstatstatements.html). Then you could see statistics on a per query basis, though I guess there is more overhead to that than this PR.

Alternatively, maybe something like [`pg_stat_io`](https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-IO-VIEW). Ee could have a `neon_stat_io`, and then someone could do something like:

```sql
SELECT lfc_hits, lfc_misses FROM neon.stat_io;
-- Run query
SELECT lfc_hits, lfc_misses FROM neon.stat_io;
-- Do math
```

Which in theory would be the same information.

Reply:
1. I definitely want to see global lfc stats.
This will allow us to estimate get_page_at_lsn() on compute size.
`neon_stats` (which I suggest to rename to `neon_lfc_stats`, btw)  is enough for a start, we can iterate on this later and turn it into view or something.

2. As for exposing this to EXPLAIN, I don't have a strong opinion here.
We won't be able to see this information, because these are user queries. Please, correct me if i'm wrong and we do expose it somewhere.
At the same time, users don't have any control over shared_buffers or LFC, so this information will only be useful when we debug queries for users that explicitly complain about performance.

Reply:
> I definitely want to see global lfc stats.

Yeh, extending standard Postgres stats seems to be really better. But it requires much more changes and what is more important - changes in Postgres core.


> We won't be able to see this information, because these are user queries. Please, correct me if i'm wrong and we do expose it somewhere.

We definitely can get access to this information if users wants: he can send us output of `explain analyze` or use `auto explain` extension to write this information in Posrtgres log which we can inspect...

Reply:
> To me, it seems useful, though I do echo what Matthias says. Here is my comment in the PG PR.
> 
> > Has Postgres ever thought about letting extensions add to EXPLAIN? Seems like that would be useful.
> 
> I am wondering if having a `neon_stat_statements` table would be good, similar to [`pg_stat_statements`](https://www.postgresql.org/docs/current/pgstatstatements.html). Then you could see statistics on a per query basis, though I guess there is more overhead to that than this PR.
> 
> Alternatively, maybe something like [`pg_stat_io`](https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-IO-VIEW). We could have a `neon_stat_io`, and then someone could do something like:
> 
> ```sql
> SELECT lfc_hits, lfc_misses FROM neon.stat_io;
> -- Run query
> SELECT lfc_hits, lfc_misses FROM neon.stat_io;
> -- Do math
> ```
> 
> Which in theory would be the same information.

Do we really need to maintain per-backend information about LFC? In case of `pg_stat_io` it is useful because some queries/session are disk intensive, some not and  such view allows to determine it. But LFC statistic is IMHP mostly interested for the whole node, it doesn't matter in which particular backend we got most cache missed.